Style check warnings:
dev:
```
/home/runner/work/gwt/gwt/gwt/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java:1121:0: warning: Line is longer than 100 characters (found 103). (com.puppycrawl.tools.checkstyle.checks.sizes.LineLengthCheck)
/home/runner/work/gwt/gwt/gwt/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java:3914:5: warning: Redundant 'public' modifier. (com.puppycrawl.tools.checkstyle.checks.modifier.RedundantModifierCheck)
```
dev test:
```
 /home/runner/work/gwt/gwt/gwt/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java:28:1: warning: Wrong order for 'org.apache.commons.lang3.StringUtils' import. (com.puppycrawl.tools.checkstyle.checks.imports.ImportOrderCheck)
/home/runner/work/gwt/gwt/gwt/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java:242:0: info: Line matches the illegal pattern 'more than one blank line'. (com.puppycrawl.tools.checkstyle.checks.regexp.RegexpCheck)
```

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -94,6 +94,7 @@ import com.google.gwt.dev.jjs.ast.JType;
 import com.google.gwt.dev.jjs.ast.JUnaryOperator;
 import com.google.gwt.dev.jjs.ast.JUnsafeTypeCoercion;
 import com.google.gwt.dev.jjs.ast.JVariable;
+import com.google.gwt.dev.jjs.ast.JVariableRef;
 import com.google.gwt.dev.jjs.ast.JWhileStatement;
 import com.google.gwt.dev.jjs.ast.js.JMultiExpression;
 import com.google.gwt.dev.jjs.ast.js.JsniClassLiteral;
@@ -1095,9 +1096,70 @@ public class GwtAstBuilder {
     public void endVisit(InstanceOfExpression x, BlockScope scope) {
       try {
         SourceInfo info = makeSourceInfo(x);
-        JExpression expr = pop(x.expression);
+        JExpression expr;
+        JDeclarationStatement jDeclarationStatement = null;
+        if (x.pattern != null) {
+          jDeclarationStatement = (JDeclarationStatement) pop();
+        }
+        expr = pop(x.expression);
         JReferenceType testType = (JReferenceType) typeMap.get(x.type.resolvedType);
-        push(new JInstanceOf(info, testType, expr));
+
+        if (jDeclarationStatement == null) {
+          push(new JInstanceOf(info, testType, expr));
+        } else {
+          // If <expr> is of type X, then
+          //
+          // rewrite (<expr> instanceof Foo foo)
+          // to
+          // Foo foo;
+          // X $instanceof_1;
+          // (($instanceof_1 = <expr>) instanceof Foo && null != (foo = (Foo) $instanceof_1))
+          //
+          // to avoid wrong behavior if o is a method call with side effects
+
+          // Foo foo;
+          String patternDeclarationName = jDeclarationStatement.getVariableRef().getTarget().getName();
+          if (!curMethod.instanceOfDeclarations.containsKey(patternDeclarationName)) {
+            curMethod.body.getBlock().addStmt(0, jDeclarationStatement);
+            curMethod.instanceOfDeclarations.put(patternDeclarationName, jDeclarationStatement);
+          }
+
+          // X $instanceof_1;
+          JType expressionType = typeMap.get(x.expression.resolvedType);
+          JLocal local =
+              createLocal(info, "$instanceOfExpr", expressionType);
+          JDeclarationStatement expressionDeclaration =
+              makeDeclaration(info, local, null);
+          curMethod.body.getBlock().addStmt(0, expressionDeclaration);
+          curMethod.instanceOfDeclarations.put(local.getName(), expressionDeclaration);
+
+          // (Foo) $instanceof_1
+          JVariableRef variableRef = jDeclarationStatement.getVariableRef();
+          JCastOperation jCastOperation =
+              new JCastOperation(info, variableRef.getType(), local.createRef(info));
+
+          // foo = (Foo) $instanceof_1
+          JBinaryOperation assignOperation =
+              new JBinaryOperation(info, variableRef.getType(), JBinaryOperator.ASG, variableRef,
+                  jCastOperation);
+
+          // null != (foo = (Foo) $instanceof_1)
+          JBinaryOperation nullCheckOperation =
+              new JBinaryOperation(info, JPrimitiveType.BOOLEAN, JBinaryOperator.NEQ,
+                  JNullLiteral.INSTANCE, assignOperation);
+
+          // $instanceof_1 = o
+          JBinaryOperation assignLocalOperation =
+              new JBinaryOperation(info, expressionType, JBinaryOperator.ASG, local.createRef(info),
+                  expr);
+
+          // (($instanceof_1 = o) instanceof Foo && null != (foo = (Foo) $instanceof_1))
+          JBinaryOperation rewrittenSwitch =
+              new JBinaryOperation(info, JPrimitiveType.BOOLEAN, JBinaryOperator.AND,
+                  new JInstanceOf(info, testType, assignLocalOperation), nullCheckOperation);
+
+          push(rewrittenSwitch);
+        }
       } catch (Throwable e) {
         throw translateException(x, e);
       }
@@ -3847,6 +3909,7 @@ public class GwtAstBuilder {
     public final Map<LocalVariableBinding, JVariable> locals = Maps.newIdentityHashMap();
     public final JMethod method;
     public final MethodScope scope;
+    public final Map<String, JStatement> instanceOfDeclarations = Maps.newHashMap();
 
     public MethodInfo(JMethod method, JMethodBody methodBody, MethodScope methodScope) {
       this.method = method;

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -1118,7 +1118,8 @@ public class GwtAstBuilder {
           // to avoid side effects from evaluating the original expression twice
 
           // Foo foo;
-          String patternDeclarationName = jDeclarationStatement.getVariableRef().getTarget().getName();
+          String patternDeclarationName = jDeclarationStatement.getVariableRef().getTarget()
+              .getName();
           if (!curMethod.instanceOfDeclarations.containsKey(patternDeclarationName)) {
             curMethod.body.getBlock().addStmt(0, jDeclarationStatement);
             curMethod.instanceOfDeclarations.put(patternDeclarationName, jDeclarationStatement);
@@ -3911,7 +3912,7 @@ public class GwtAstBuilder {
     public final MethodScope scope;
     public final Map<String, JStatement> instanceOfDeclarations = Maps.newHashMap();
 
-    public MethodInfo(JMethod method, JMethodBody methodBody, MethodScope methodScope) {
+    MethodInfo(JMethod method, JMethodBody methodBody, MethodScope methodScope) {
       this.method = method;
       this.body = methodBody;
       this.scope = methodScope;

--- a/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
+++ b/dev/core/src/com/google/gwt/dev/jjs/impl/GwtAstBuilder.java
@@ -1115,7 +1115,7 @@ public class GwtAstBuilder {
           // X $instanceof_1;
           // (($instanceof_1 = <expr>) instanceof Foo && null != (foo = (Foo) $instanceof_1))
           //
-          // to avoid wrong behavior if o is a method call with side effects
+          // to avoid side effects from evaluating the original expression twice
 
           // Foo foo;
           String patternDeclarationName = jDeclarationStatement.getVariableRef().getTarget().getName();

--- a/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java
+++ b/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java
@@ -16,12 +16,16 @@ package com.google.gwt.dev.jjs.impl;
 import com.google.gwt.core.ext.UnableToCompleteException;
 import com.google.gwt.dev.javac.testing.impl.JavaResourceBase;
 import com.google.gwt.dev.jjs.InternalCompilerException;
+import com.google.gwt.dev.jjs.ast.JAbstractMethodBody;
+import com.google.gwt.dev.jjs.ast.JDeclaredType;
 import com.google.gwt.dev.jjs.ast.JInterfaceType;
 import com.google.gwt.dev.jjs.ast.JProgram;
 import com.google.gwt.dev.jjs.ast.JStringLiteral;
+import com.google.gwt.dev.jjs.ast.JType;
 
 import java.util.Arrays;
 import java.util.List;
+import org.apache.commons.lang3.StringUtils;
 
 /**
  * Tests that {@link GwtAstBuilder} correctly builds the AST for
@@ -46,18 +50,33 @@ public class Java17AstTest extends FullCompileTestBase {
     addAll(JavaResourceBase.createMockJavaResource("test.Shape",
         "package test;",
         "public sealed class Shape permits Square, Circle {",
+        "private static int count = 0;",
+        "public static Shape returnAndIncrement(Shape shape) {",
+        "count++;",
+        "return shape;",
+        "}",
         "}"
     ));
 
     addAll(JavaResourceBase.createMockJavaResource("test.Square",
         "package test;",
         "public final class Square extends Shape {",
+
+        "public int getLength() {",
+        "return 10;",
+        "}",
+        "public double getSide() {",
+        "return 0;",
+        "}",
         "}"
     ));
 
     addAll(JavaResourceBase.createMockJavaResource("test.Circle",
         "package test;",
         "public final class Circle extends Shape {",
+        "public double getDiameter() {",
+        "return 0;",
+        "}",
         "}"
     ));
 
@@ -65,6 +84,20 @@ public class Java17AstTest extends FullCompileTestBase {
         "package test;",
         "public enum Months {",
         "JANUARY, FEBRUARY, MARCH, APRIL, MAY, JUNE, JULY, AUGUST, SEPTEMBER, OCTOBER, NOVEMBER, DECEMBER;",
+        "}"
+    ));
+
+    addAll(JavaResourceBase.createMockJavaResource("test.TestSupplier",
+        "package test;",
+        "public interface TestSupplier {",
+        "  boolean run();",
+        "}"
+    ));
+
+    addAll(JavaResourceBase.createMockJavaResource("test.Polygon",
+        "package test;",
+        "public class Polygon {",
+        "  public int sides;",
         "}"
     ));
   }
@@ -135,6 +168,95 @@ public class Java17AstTest extends FullCompileTestBase {
     }
   }
 
+  public void testInstanceOfPatternMatching() throws UnableToCompleteException {
+    JProgram program = compileSnippet("void", "Shape shape1 = new Circle();" +
+        "if(shape1 instanceof Circle circle) {" +
+        "double diameter = circle.getDiameter();" +
+        "}"
+    );
+
+    JAbstractMethodBody onModuleLoadMethod = findMethod(program, "onModuleLoad")
+        .getBody();
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Circle circle;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_3;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_3 = shape1) instanceof Circle && null != (circle = (Circle) $instanceOfExpr_3)"));
+  }
+
+  public void testInstanceOfPatternMatchingWithSideEffectsExpression()
+      throws UnableToCompleteException {
+    JProgram program = compileSnippet("void", "Shape shape1 = new Circle();" +
+        "if(Shape.returnAndIncrement(shape1) instanceof Circle circle) {" +
+        "double diameter = circle.getDiameter();" +
+        "}"
+    );
+
+    JAbstractMethodBody onModuleLoadMethod = findMethod(program, "onModuleLoad")
+        .getBody();
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Circle circle;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_3;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_3 = Shape.returnAndIncrement(shape1)) instanceof Circle && null != (circle = (Circle) $instanceOfExpr_3)"));
+  }
+
+  public void testInstanceOfPatternMatchingWithAnd() throws UnableToCompleteException {
+    JProgram program = compileSnippet("void",
+        "Shape shape1 = new Circle();\n" +
+        "Shape shape2 = new Square();\n" +
+        "if(shape1 instanceof Circle circle && shape2 instanceof Square square) {\n" +
+        "double diameter = circle.getDiameter();\n" +
+        "}\n"
+    );
+
+    JAbstractMethodBody onModuleLoadMethod = findMethod(program, "onModuleLoad")
+        .getBody();
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Circle circle;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_4;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Square square;"));
+
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_6"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_4 = shape1) instanceof Circle && null != (circle = (Circle) $instanceOfExpr_4)"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_6 = shape2) instanceof Square && null != (square = (Square) $instanceOfExpr_6)"));
+  }
+
   public void testSwitchExpressionsInitializerShouldFail() {
     try {
       compileSnippet("void", "    int i = switch(1) {\n" +
@@ -152,6 +274,157 @@ public class Java17AstTest extends FullCompileTestBase {
       }
       assertEquals("Switch expressions not yet supported", e.getMessage());
     }
+  }
+
+  public void testInstanceOfPatternMatchingWithCondition() throws UnableToCompleteException {
+    JProgram program = compileSnippet("void",
+        "Shape shape2 = new Square();\n" +
+        "if(shape2 instanceof Square square && square.getLength() > 0) {\n" +
+        "double diameter = square.getSide();\n" +
+        "}\n"
+    );
+
+    JAbstractMethodBody onModuleLoadMethod = findMethod(program, "onModuleLoad")
+        .getBody();
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Square square;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_3;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_3 = shape2) instanceof Square && null != (square = (Square) $instanceOfExpr_3)"));
+  }
+
+  public void testInstanceOfPatternMatchingWithAsNotCondition() throws UnableToCompleteException {
+    JProgram program = compileSnippet("void",
+        "Shape shape1 = new Square();\n" +
+        "if(!(shape1 instanceof Square square && square.getLength() > 0)) {\n" +
+        "}\n"
+    );
+
+    JAbstractMethodBody onModuleLoadMethod = findMethod(program, "onModuleLoad")
+        .getBody();
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Square square;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_3;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_3 = shape1) instanceof Square && null != (square = (Square) $instanceOfExpr_3)"));
+  }
+
+  public void testMultipleInstanceOfPatternMatchingWithSameVariableName() throws UnableToCompleteException {
+    JProgram program = compileSnippet("void",
+        "Shape shape1 = new Square();\n" +
+        "Shape shape2 = new Square();\n" +
+        "if(shape1 instanceof Square square && square.getLength() > 0) {\n" +
+        "}\n" +
+        "if(shape2 instanceof Square square && square.getLength() > 0) {\n" +
+        "}\n"
+    );
+
+    JAbstractMethodBody onModuleLoadMethod = findMethod(program, "onModuleLoad")
+        .getBody();
+
+    assertEquals(1, StringUtils.countMatches(onModuleLoadMethod.toSource(), "Square square;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_4;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains("Shape $instanceOfExpr_6;"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_4 = shape1) instanceof Square && null != (square = (Square) $instanceOfExpr_4)"));
+
+    assertTrue(onModuleLoadMethod
+        .toSource()
+        .contains(
+            "($instanceOfExpr_6 = shape2) instanceof Square && null != (square = (Square) $instanceOfExpr_6)"));
+  }
+
+  public void testInstanceOfPatternMatchingInLambda() throws UnableToCompleteException {
+    addSnippetClassDecl("public class Foo {\n" +
+        "private Shape shape;\n" +
+        "public Foo(){\n" +
+        "shape = new Square();\n" +
+        "}\n" +
+        "public TestSupplier isSquare(){\n" +
+        "return () -> shape instanceof Square square && square.getLength() > 0;\n" +
+        "}\n" +
+        "}");
+
+    JProgram program = compileSnippet("void", "Foo foo = new Foo();");
+    JType foo = findType(program, "test.EntryPoint.Foo");
+    JAbstractMethodBody lambda = findMethod((JDeclaredType) foo, "lambda$0")
+        .getBody();
+
+    assertTrue(lambda
+        .toSource()
+        .contains("Square square;"));
+
+    assertTrue(lambda
+        .toSource()
+        .contains("Shape $instanceOfExpr_2;"));
+
+    assertTrue(lambda
+        .toSource()
+        .contains(
+            "($instanceOfExpr_2 = this.shape) instanceof Square && null != (square = (Square) $instanceOfExpr_2)"));
+  }
+
+  public void testInstanceOfPatternMatchingAsReturn() throws UnableToCompleteException {
+
+    addSnippetClassDecl("public class Foo {\n" +
+        "private Shape shape;\n" +
+        "public Foo(){\n" +
+        "shape = new Square();\n" +
+        "}\n" +
+        "public boolean isSquare(){\n" +
+        "return shape instanceof Square square && square.getLength() > 0;\n" +
+        "}\n" +
+        "}");
+    JProgram program = compileSnippet("void", "Foo foo = new Foo();");
+    JType foo = findType(program, "test.EntryPoint.Foo");
+    JAbstractMethodBody isSquare = findMethod((JDeclaredType) foo, "isSquare")
+        .getBody();
+
+    assertTrue(isSquare
+        .toSource()
+        .contains("Square square;"));
+
+    assertTrue(isSquare
+        .toSource()
+        .contains("Shape $instanceOfExpr_2;"));
+
+    assertTrue(isSquare
+        .toSource()
+        .contains(
+            "($instanceOfExpr_2 = this.shape) instanceof Square && null != (square = (Square) $instanceOfExpr_2)"));
+  }
+
+  public void testInstanceOfPatternMatchingWithConditionalOperator()
+      throws UnableToCompleteException {
+      compileSnippet("void", "var a = true ? \"a\":'a';\n" +
+          "if(a instanceof CharSequence p) {\n" +
+          "     int b= p.length();\n" +
+          "}");
   }
 
   @Override

--- a/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java
+++ b/dev/core/test/com/google/gwt/dev/jjs/impl/Java17AstTest.java
@@ -23,9 +23,10 @@ import com.google.gwt.dev.jjs.ast.JProgram;
 import com.google.gwt.dev.jjs.ast.JStringLiteral;
 import com.google.gwt.dev.jjs.ast.JType;
 
+import org.apache.commons.lang3.StringUtils;
+
 import java.util.Arrays;
 import java.util.List;
-import org.apache.commons.lang3.StringUtils;
 
 /**
  * Tests that {@link GwtAstBuilder} correctly builds the AST for
@@ -240,7 +241,6 @@ public class Java17AstTest extends FullCompileTestBase {
     assertTrue(onModuleLoadMethod
         .toSource()
         .contains("Square square;"));
-
 
     assertTrue(onModuleLoadMethod
         .toSource()

--- a/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
+++ b/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
@@ -41,17 +41,54 @@ public class Java17Test extends GWTTestCase {
   }
 
   public sealed class Shape permits Square, Circle {
+    public static int count = 0;
+
+    public static Shape returnAndIncrement(Shape shape){
+      count++;
+      return shape;
+    }
 
   }
 
   public final class Square extends Shape {
-
+    public int getLength() {
+      return 10;
+    }
   }
 
   public final class Circle extends Shape {
-
+    public int getDiameter() {
+      return 10;
+    }
   }
 
+  public interface TestSupplier {
+    boolean run();
+  }
+
+  public class Foo {
+    private Shape shape;
+
+    public Foo() {
+      shape = new Square();
+    }
+
+    public TestSupplier isSquare() {
+      return () -> shape instanceof Square square && square.getLength() > 0;
+    }
+  }
+
+  public class Bar {
+    private Shape shape;
+
+    public Bar() {
+      shape = new Square();
+    }
+
+    public boolean isSquare() {
+      return shape instanceof Square square && square.getLength() > 0;
+    }
+  }
   @Override
   public String getModuleName() {
     return "com.google.gwt.dev.jjs.test.Java17Test";
@@ -75,5 +112,104 @@ public class Java17Test extends GWTTestCase {
   private void checkIfCompiled(Shape square, Shape circle) {
     assertTrue(square instanceof Square);
     assertTrue(circle instanceof Circle);
+  }
+
+  public void testInstanceOfPatternMatching() {
+    Shape shape1 = new Circle();
+    if (shape1 instanceof Circle circle) {
+      circle.getDiameter();
+      assertTrue(true);
+      return;
+    }
+    fail();
+  }
+
+  public void testInstanceOfPatternMatchingWithSideEffectExpression() {
+    Shape shape1 = new Circle();
+    if (Shape.returnAndIncrement(shape1) instanceof Circle circle) {
+      circle.getDiameter();
+      assertTrue(true);
+      assertEquals(1, Shape.count);
+      return;
+    }
+    fail();
+  }
+
+  public void testInstanceOfPatternMatchingWithAnd() {
+    Shape shape1 = new Circle();
+    Shape shape2 = new Square();
+
+    if (shape1 instanceof Circle circle && shape2 instanceof Square square) {
+      circle.getDiameter();
+      square.getLength();
+      assertTrue(true);
+      return;
+    }
+    fail();
+  }
+
+  public void testInstanceOfPatternMatchingWithCondition() {
+    Shape shape2 = new Square();
+    if (shape2 instanceof Square square && square.getLength() > 0) {
+      square.getLength();
+      assertTrue(true);
+      return;
+    }
+    fail();
+  }
+
+  public void testInstanceOfPatternMatchingWithAsNotCondition() {
+    Shape shape1 = new Square();
+    if (!(shape1 instanceof Square square && square.getLength() > 10)) {
+      assertTrue(true);
+      return;
+    }
+    fail();
+  }
+
+  public void testMultipleInstanceOfPatternMatchingWithSameVariableName() {
+    Shape shape1 = new Square();
+    Shape shape2 = new Square();
+    boolean a = false;
+    boolean b = false;
+    if (shape1 instanceof Square square && square.getLength() > 0) {
+      a = true;
+    }
+    if (shape2 instanceof Square square && square.getLength() > 0) {
+      b = true;
+    }
+    assertTrue(a && b);
+  }
+
+  public void testMultipleInstanceOfPatternMatchingWithSameVariableNameWithDifferentTypes() {
+    Shape shape1 = new Square();
+    Shape shape2 = new Circle();
+    boolean a = false;
+    boolean b = false;
+    if (shape1 instanceof Square shp && shp.getLength() > 0) {
+      a = true;
+    }
+    if (shape2 instanceof Circle shp && shp.getLength() > 0) {
+      b = true;
+    }
+    assertTrue(a && b);
+  }
+
+  public void testMultipleInstanceOfPatternMatchingWithSameVariableNameWithDifferentTypes() {
+    Shape shape1 = new Square();
+    if (shape1 instanceof Circle shp) {
+      fail("Should have not reached this point.");
+    }
+    assertTrue(true);
+  }
+
+  public void testInstanceOfPatternMatchingInLambda() {
+    Foo foo = new Foo();
+    assertTrue(foo.isSquare().run());
+  }
+
+  public void testInstanceOfPatternMatchingAsReturn() {
+    Bar bar = new Bar();
+    assertTrue(bar.isSquare());
   }
 }

--- a/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
+++ b/user/test-super/com/google/gwt/dev/jjs/super/com/google/gwt/dev/jjs/test/Java17Test.java
@@ -189,13 +189,13 @@ public class Java17Test extends GWTTestCase {
     if (shape1 instanceof Square shp && shp.getLength() > 0) {
       a = true;
     }
-    if (shape2 instanceof Circle shp && shp.getLength() > 0) {
+    if (shape2 instanceof Circle shp && shp.getDiameter() > 0) {
       b = true;
     }
     assertTrue(a && b);
   }
 
-  public void testMultipleInstanceOfPatternMatchingWithSameVariableNameWithDifferentTypes() {
+  public void testInstanceOfPatternMatchingIsFalse() {
     Shape shape1 = new Square();
     if (shape1 instanceof Circle shp) {
       fail("Should have not reached this point.");

--- a/user/test/com/google/gwt/dev/jjs/test/Java17Test.java
+++ b/user/test/com/google/gwt/dev/jjs/test/Java17Test.java
@@ -50,6 +50,34 @@ public class Java17Test extends GWTTestCase {
     assertFalse(isGwtSourceLevel17());
   }
 
+  public void testInstanceOfPatternMatching() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
+  public void testInstanceOfPatternMatchingWithAnd() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
+  public void testInstanceOfPatternMatchingWithCondition() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
+  public void testInstanceOfPatternMatchingWithAsNotCondition() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
+  public void testMultipleInstanceOfPatternMatchingWithSameVariableName() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
+  public void testInstanceOfPatternMatchingInLambda() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
+  public void testInstanceOfPatternMatchingAsReturn() {
+    assertFalse(isGwtSourceLevel17());
+  }
+
   private boolean isGwtSourceLevel17() {
     return JUnitShell.getCompilerOptions().getSourceLevel().compareTo(SourceLevel.JAVA17) >= 0;
   }

